psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -888,24 +977,452 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+                        found = true;
+                        atom_pairs.emplace_back(Patom, Qatom);
+                        break;
+                    }
+                }
+                if (found) break;
+            }
+        }
+    }
+
+    // A comparator used for sorting integral screening values
+    auto screen_compare = [](const std::pair<int, double> &a, 
+                                    const std::pair<int, double> &b) { return a.second > b.second; };
+
+    // Shells linked to each other through Schwarz Screening (Significant Overlap)
+    std::vector<std::vector<int>> significant_bras(nshell);
+    double max_integral = ints[0]->max_integral();
+
+    for (size_t P = 0; P < nshell; P++) {
+        std::vector<std::pair<int, double>> PQ_shell_values;
+        for (size_t Q = 0; Q < nshell; Q++) {
+            double pq_pq = std::sqrt(ints[0]->shell_ceiling2(P, Q, P, Q));
+            double schwarz_value = std::sqrt(pq_pq * max_integral);
+            if (schwarz_value >= cutoff_) {
+                PQ_shell_values.emplace_back(Q, schwarz_value);
+            }
+        }
+        std::sort(PQ_shell_values.begin(), PQ_shell_values.end(), screen_compare);
+
+        for (const auto& value : PQ_shell_values) {
+            significant_bras[P].push_back(value.first);
+        }
+    }
+
+    // => Calculate Shell Ceilings (To find significant bra-ket pairs)
+    // sqrt(Umax|Umax) in Oschenfeld Eq. 3
+    std::vector<double> shell_ceilings(nshell, 0.0);
+    for (int P = 0; P < nshell; P++) {
+        for (int Q = 0; Q <= P; Q++) {
+            double val = std::sqrt(ints[0]->shell_ceiling2(P, Q, P, Q));
+            shell_ceilings[P] = std::max(shell_ceilings[P], val);
+            shell_ceilings[Q] = std::max(shell_ceilings[Q], val);
+        }
+    }
+
+    size_t natom_pair = atom_pairs.size();
+
+    // => Intermediate Buffers <= //
+
+    // Temporary buffers used during the K contraction process to
+    // Take full advantage of permutational symmetry of ERIs
+    std::vector<std::vector<SharedMatrix>> KT;
+
+    // A buffer is created for every thread to minimize race conditions
+    for (int thread = 0; thread < nthread; thread++) {
+        std::vector<SharedMatrix> K2;
+        for (size_t ind = 0; ind < D.size(); ind++) {
+            // (pq|rs) can be contracted into Kpr, Kps, Kqr, Kqs (hence the 4)
+            K2.push_back(std::make_shared<Matrix>("KT (linK)", 4 * max_functions_per_atom, nbf));
+        }
+        KT.push_back(K2);
+    }
+
+    // Number of computed shell quartets is tracked for benchmarking purposes
+    size_t computed_shells = 0L;
+
+// ==> Master Task Loop (Atom Quartet Indexing) <== //
+
+// ==> "Loop over types (angular momenta, contraction, ...) of shell-pair blocks" <== //
+#pragma omp parallel for num_threads(nthread) schedule(dynamic) reduction(+ : computed_shells)
+    for (size_t ipair = 0L; ipair < natom_pair; ipair++) { // O(N) shell-pairs in asymptotic limit
```

Comment:
Can we use a range-based for loop over `atom_pairs` here?

Reply:
I do not think range-based for loops work with OpenMP